### PR TITLE
Crazy Refactoring

### PR DESCRIPTION
**Pull request by chaosmail**
here is a PR, that you can see on what I am working on.

I have loads of remarks and comments on testing, but this would fill another hangout session. Maybe we can do this next week, so we can agree on some testing guidelines.

Best.
Christoph

### CODE REVIEW

**Comment by chaosmail** [comment]
yeah, and I removed the `toJS` methods. I mean basically we are dealing with TypeScript objects - which are ES5 objects. So there is no need for these functions I think

**Comment by chaosmail** [comment]
I want to propose the following for tests: 
* every unit test tests exactly one function
* we define a `expected` and a `testing` variable in every test (only exception if we are testing if a function throws an error)
* we use a describe block per function to test
* common code goes into the beforeEach block in this describe function
* we should not need testUtils
* everything that needs more than one function should be an e2e test (such as events, etc.)

**Comment by lorem--ipsum** [comment]
I totally agree with your ideas, let's write that in the wiki !

File: src/LineChart.ts [hunk 1]
```diff
@@ -66,12 +70,20 @@ module n3Charts {
         eventMgr.trigger('update', data, options);
       }, true);
 
+      scope.$watch('onDatumEnter', function() {
```
Comment (lorem--ipsum) [inline src/LineChart.ts]: Maybe we could avoid repeating the same chunk of code with different event name...

Comment (chaosmail) [inline src/LineChart.ts]: totally agree

**Comment by lorem--ipsum** [comment]
Okay I just read your code, it's great. Still not looking for a full time job ? ^^
Do you want to work on my only comment or should I merge the PR ?

**Comment by chaosmail** [comment]
I think we can merge this, to have the latest version to work on.

**Comment by lorem--ipsum** [comment]
:poultry_leg:

**Pull request merged**